core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java

```diff
@@ -1271,7 +1271,7 @@ protected final void check(
     sql(sql).expand(false).ok();
   }
 
-  /** Since 'deptno NOT IN (SELECT deptno FROM dept)' can not be null, we
+  /** Since 'deptno NOT IN (SELECT deptno FROM dept)' can be null, we
```

Comment:
Why did you add "not"? Neither emp.deptno nor dept.deptno can be null, therefore NOT IN can never be unknown. Which is very useful, because we can generate a more efficient plan.

Reply:
I removed it because, without additional changes, the subquery becomes a left outer join and therefore the fields are nullable.  With this current patch, NOT(IS_TRUE(X)) does not get reduced to IS_FALSE(X) if X is nullable.  I will make some additional changes to make sure that this optimization is preserved.

Reply:
Sure, it's implemented as a LEFT join, but that gives 2 possibilities - exists and does not exist - which is manifested as a check as to whether a TRUE column is still TRUE or has turned into NULL due to the outer join. That's a lot better than 3 possibilities, which is the case that NOT IN usually has to handle. Optimizing to 2 possibilities is exactly what this test is checking for. It is possible precisely because emp.deptno and dept.deptno are provably never null in the query.